File: frontend/src/lib/patients.ts
import { post, request } from "@/lib/api";

export type Gender = "female" | "male" | "other";
export type BloodGroup = "A+" | "A-" | "B+" | "B-" | "AB+" | "AB-" | "O+" | "O-";
export type Severity = "mild" | "moderate" | "severe";

export const GENDERS: Gender[] = ["female", "male", "other"];
export const BLOOD_GROUPS: BloodGroup[] = ["A+", "A-", "B+", "B-", "AB+", "AB-", "O+", "O-"];
export const SEVERITIES: Severity[] = ["mild", "moderate", "severe"];

export type PatientAddress = {
  line1: string;
  line2: string;
  city: string;
  state: string;
  postal_code: string;
  country: string;
};

export type EmergencyContact = {
  name: string;
  relationship: string;
  phone: string;
};

export type Allergy = {
  id: string;
  substance: string;
  reaction: string | null;
  severity: Severity;
  created_at: string;
};

export type PatientSummary = {
  id: string;
  patient_number: string;
  full_name: string;
  preferred_name: string | null;
  phone: string | null;
  date_of_birth: string | null;
  age: string | null;
  gender: Gender | null;
  blood_group: BloodGroup | null;
  status: "active" | "archived";
  allergy_count: number;
  created_at: string;
};

export type Patient = PatientSummary & {
  first_name: string;
  last_name: string;
  alternate_phone: string | null;
  email: string | null;
  address: Partial<PatientAddress> | null;
  emergency_contact: Partial<EmergencyContact> | null;
  notes: string | null;
  archived_at: string | null;
  registered_by_name: string | null;
  allergies: Allergy[];
};

/** A record that might be the same person, and the reason it surfaced. */
export type DuplicateCandidate = PatientSummary & { reason: string };

export type PatientPage = {
  items: PatientSummary[];
  total: number;
  page: number;
  per_page: number;
  pages: number;
};

export type PatientDraft = {
  first_name: string;
  last_name: string;
  preferred_name?: string | null;
  phone?: string | null;
  alternate_phone?: string | null;
  email?: string | null;
  date_of_birth?: string | null;
  gender?: Gender | null;
  blood_group?: BloodGroup | null;
  address?: Partial<PatientAddress> | null;
  emergency_contact?: Partial<EmergencyContact> | null;
  notes?: string | null;
};

export type PatientQuery = {
  q?: string;
  status?: "active" | "archived" | "all";
  page?: number;
  per_page?: number;
};

export const PAGE_SIZE = 25;

export function listPatients({
  q = "",
  status = "active",
  page = 1,
  per_page = PAGE_SIZE,
}: PatientQuery = {}): Promise<PatientPage> {
  const params = new URLSearchParams({
    status,
    page: String(page),
    per_page: String(per_page),
  });
  if (q.trim()) params.set("q", q.trim());
  return request<PatientPage>(`/patients?${params}`);
}

export const getPatient = (id: string) => request<Patient>(`/patients/${id}`);

export const registerPatient = (draft: PatientDraft & { confirm_duplicate?: boolean }) =>
  post<Patient>("/patients", draft);

export const savePatient = (id: string, changes: Partial<PatientDraft>) =>
  request<Patient>(`/patients/${id}`, { method: "PATCH", body: JSON.stringify(changes) });

export const archivePatient = (id: string) => post<Patient>(`/patients/${id}/archive`, {});

export const restorePatient = (id: string) => post<Patient>(`/patients/${id}/restore`, {});

export const checkDuplicates = (draft: {
  first_name?: string;
  last_name?: string;
  phone?: string | null;
  email?: string | null;
  date_of_birth?: string | null;
  exclude_id?: string;
}) => post<DuplicateCandidate[]>("/patients/check-duplicates", draft);

export const recordAllergy = (
  patientId: string,
  allergy: { substance: string; reaction?: string | null; severity: Severity },
) => post<Allergy>(`/patients/${patientId}/allergies`, allergy);

export const removeAllergy = (patientId: string, allergyId: string) =>
  request<{ removed: boolean }>(`/patients/${patientId}/allergies/${allergyId}`, {
    method: "DELETE",
  });

/** Initials for the avatar on the list, from whichever names exist. */
export function initials(name: string): string {
  const parts = name.trim().split(/\s+/).filter(Boolean);
  const first = parts.at(0);
  if (!first) return "?";
  const last = parts.at(-1);
  if (parts.length === 1) return first.slice(0, 2).toUpperCase();
  return `${first.charAt(0)}${last?.charAt(0) ?? ""}`.toUpperCase();
}

/** A phone number grouped the way it is read back, without changing it. */
export function readablePhone(phone: string | null): string {
  if (!phone) return "";
  const local = phone.startsWith("+91") ? phone.slice(3) : phone;
  if (local.length === 10) {
    const spaced = `${local.slice(0, 5)} ${local.slice(5)}`;
    return phone.startsWith("+91") ? `+91 ${spaced}` : spaced;
  }
  return phone;
}

/** A line describing a person at a glance: "38 y · female · O+". */
export function describe(patient: PatientSummary): string {
  return [patient.age, patient.gender, patient.blood_group].filter(Boolean).join(" · ");
}

File: backend/app/api/v1/endpoints/patients.py
"""Patient records.

Every route resolves the caller and is gated on a permission, and every read
goes through the scoped repository, so a record belonging to another clinic
is not found rather than forbidden.
"""

from __future__ import annotations

import math
import uuid

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import Caller, DbSession, current_tenant, requires
from app.models.patient import ACTIVE
from app.repositories.patients import AllergyRepository, Listed, PatientRepository, names_of
from app.schemas.patient import (
    AllergyOut,
    AllergyWrite,
    DuplicateCandidate,
    DuplicateCheck,
    PatientCreate,
    PatientOut,
    PatientPage,
    PatientSummary,
    PatientUpdate,
    Removed,
)
from app.services import patients as service

router = APIRouter(tags=["patients"])


def _summary(row: Listed) -> PatientSummary:
    patient = row.patient
    return PatientSummary(
        id=patient.id,
        patient_number=patient.patient_number,
        full_name=patient.full_name,
        preferred_name=patient.preferred_name,
        phone=patient.phone,
        date_of_birth=patient.date_of_birth,
        age=service.age_label(patient.date_of_birth),
        gender=patient.gender,
        blood_group=patient.blood_group,
        status=patient.status,
        allergy_count=row.allergy_count,
        created_at=patient.created_at,
    )


def _detail(
    row: Listed, allergies: list[AllergyOut], registered_by_name: str | None
) -> PatientOut:
    patient = row.patient
    return PatientOut(
        **_summary(row).model_dump(),
        first_name=patient.first_name,
        last_name=patient.last_name,
        alternate_phone=patient.alternate_phone,
        email=patient.email,
        address=patient.address,
        emergency_contact=patient.emergency_contact,
        notes=patient.notes,
        archived_at=patient.archived_at,
        registered_by_name=registered_by_name,
        allergies=allergies,
    )


async def _read(
    session: AsyncSession, organization_id: uuid.UUID, patient_id: uuid.UUID
) -> PatientOut:
    row = await service.fetch(session, organization_id=organization_id, patient_id=patient_id)
    allergies = await AllergyRepository(session, organization_id).for_patient(patient_id)

    registered_by = row.patient.registered_by_id
    names = await names_of(session, {registered_by} if registered_by else set())

    return _detail(
        row,
        [AllergyOut.model_validate(allergy) for allergy in allergies],
        names.get(registered_by) if registered_by else None,
    )


@router.get("/patients")
async def list_patients(
    session: DbSession,
    query: str = Query(default="", max_length=120, alias="q"),
    status: str = Query(default=ACTIVE, pattern="^(active|archived|all)$"),
    page: int = Query(default=1, ge=1),
    per_page: int = Query(default=25, ge=1, le=100),
    _: Caller = Depends(requires("patient:read")),
    organization_id: uuid.UUID = Depends(current_tenant),
) -> PatientPage:
    found, total = await PatientRepository(session, organization_id).search(
        query=query,
        status=None if status == "all" else status,
        limit=per_page,
        offset=(page - 1) * per_page,
    )
    return PatientPage(
        items=[_summary(row) for row in found],
        total=total,
        page=page,
        per_page=per_page,
        # Carried so a client can offer a way forward without a second
        # request to discover whether there is one.
        pages=max(math.ceil(total / per_page), 1),
    )


@router.post("/patients/check-duplicates")
async def check_duplicates(
    session: DbSession,
    body: DuplicateCheck,
    _: Caller = Depends(requires("patient:read")),
    organization_id: uuid.UUID = Depends(current_tenant),
) -> list[DuplicateCandidate]:
    """Asked while the form is still being filled in, so whoever is at the
    desk sees the existing record in time to open it instead."""
    found = await PatientRepository(session, organization_id).possible_duplicates(
        first_name=body.first_name,
        last_name=body.last_name,
        phone=body.phone,
        email=body.email,
        date_of_birth=body.date_of_birth,
        exclude_id=body.exclude_id,
    )
    return [
        DuplicateCandidate(
            **_summary(row).model_dump(),
            reason=service.candidate_reason(row.patient, phone=body.phone, email=body.email),
        )
        for row in found
    ]


@router.post("/patients", status_code=201)
async def register_patient(
    session: DbSession,
    body: PatientCreate,
    caller: Caller = Depends(requires("patient:create")),
    organization_id: uuid.UUID = Depends(current_tenant),
) -> PatientOut:
    patient = await service.register(
        session, organization_id=organization_id, actor_id=caller.user_id, body=body
    )
    return _detail(Listed(patient=patient, allergy_count=0), [], None)


@router.get("/patients/{patient_id}")
async def read_patient(
    session: DbSession,
    patient_id: uuid.UUID,
    _: Caller = Depends(requires("patient:read")),
    organization_id: uuid.UUID = Depends(current_tenant),
) -> PatientOut:
    return await _read(session, organization_id, patient_id)


@router.patch("/patients/{patient_id}")
async def update_patient(
    session: DbSession,
    patient_id: uuid.UUID,
    body: PatientUpdate,
    _: Caller = Depends(requires("patient:update")),
    organization_id: uuid.UUID = Depends(current_tenant),
) -> PatientOut:
    await service.update(
        session, organization_id=organization_id, patient_id=patient_id, body=body
    )
    return await _read(session, organization_id, patient_id)


@router.post("/patients/{patient_id}/archive")
async def archive_patient(
    session: DbSession,
    patient_id: uuid.UUID,
    _: Caller = Depends(requires("patient:archive")),
    organization_id: uuid.UUID = Depends(current_tenant),
) -> PatientOut:
    """Hides the record from the day-to-day lists. Nothing is deleted: the
    visits, invoices and notes that hang off it stay exactly as they were."""
    await service.set_archived(
        session, organization_id=organization_id, patient_id=patient_id, archived=True
    )
    return await _read(session, organization_id, patient_id)


@router.post("/patients/{patient_id}/restore")
async def restore_patient(
    session: DbSession,
    patient_id: uuid.UUID,
    _: Caller = Depends(requires("patient:archive")),
    organization_id: uuid.UUID = Depends(current_tenant),
) -> PatientOut:
    await service.set_archived(
        session, organization_id=organization_id, patient_id=patient_id, archived=False
    )
    return await _read(session, organization_id, patient_id)


@router.post("/patients/{patient_id}/allergies", status_code=201)
async def record_allergy(
    session: DbSession,
    patient_id: uuid.UUID,
    body: AllergyWrite,
    caller: Caller = Depends(requires("patient:update")),
    organization_id: uuid.UUID = Depends(current_tenant),
) -> AllergyOut:
    allergy = await service.add_allergy(
        session,
        organization_id=organization_id,
        patient_id=patient_id,
        actor_id=caller.user_id,
        body=body,
    )
    return AllergyOut.model_validate(allergy)


@router.delete("/patients/{patient_id}/allergies/{allergy_id}")
async def remove_allergy(
    session: DbSession,
    patient_id: uuid.UUID,
    allergy_id: uuid.UUID,
    _: Caller = Depends(requires("patient:update")),
    organization_id: uuid.UUID = Depends(current_tenant),
) -> Removed:
    await service.remove_allergy(
        session,
        organization_id=organization_id,
        patient_id=patient_id,
        allergy_id=allergy_id,
    )
    return Removed()
